### PR TITLE
terrain: skirting

### PR DESCRIPTION
**Pull request by belom88**
https://user-images.githubusercontent.com/18549629/125440749-616bfb33-9430-4e1d-b362-8184956e94bd.mp4

### CODE REVIEW

File: modules/terrain/docs/api-reference/quantized-mesh-loader.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ const data = await load(url, QuantizedMeshLoader, options);
 | Option                  | Type            | Default        | Description                                                                     |
 | ----------------------- | --------------- | -------------- | ------------------------------------------------------------------------------- |
 | `quantized-mesh.bounds` | `array<number>` | `[0, 0, 1, 1]` | Bounds of the image to fit x,y coordinates into. In `[minX, minY, maxX, maxY]`. |
+| `terrain.skirtHeight`   | `number`        | `null`         | If set, create the skirt for the tile with particular height                    |
```
Comment (kylebarron) [inline modules/terrain/docs/api-reference/quantized-mesh-loader.md:42]: We should define what units this height is in. I assume it's in meters if the elevation values are in meters?

File: modules/terrain/docs/api-reference/terrain-loader.md [hunk 2]
```diff
@@ -33,6 +33,7 @@ const data = await load(url, TerrainLoader, options);
 | `terrain.bounds`           | `array<number>` | `null`    | Bounds of the image to fit x,y coordinates into. In `[minX, minY, maxX, maxY]`. If not supplied, x and y are in pixels relative to the image. |
 | `terrain.elevationDecoder` | `object`        | See below | See below                                                                                                                                     |
 | `terrain.tesselator`       | `string`        | `auto`    | See below                                                                                                                                     |
+| `terrain.skirtHeight`      | `number`        | `null`    | If set, create the skirt for the tile with particular height                                                                                  |
```
Comment (kylebarron) [inline modules/terrain/docs/api-reference/terrain-loader.md:36]: Ditto.

File: modules/terrain/src/lib/helpers/skirt.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+/**
+ * Add skirt to existing mesh
+ * @param {object} attributes - POSITION and TEXCOOD_0 attributes data
+ * @param {any} triangles - indices array of the mesh geometry
+ * @param {number} skirtHeight - height of the skirt geometry
+ * @returns - geometry data with added skirt
+ */
+export function addSkirt(attributes, triangles, skirtHeight) {
+  const outsideEdges = getOutsideEdges(triangles);
```
Comment (kylebarron) [inline modules/terrain/src/lib/helpers/skirt.js:11]: This should be an optional argument to the `addSkirt` function. The quantized mesh metadata already contains this information, so you don't need to compute it. (Martini and Delatin output from the `TerrainLoader` doesn't compute this automatically)

File: modules/terrain/src/lib/helpers/skirt.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
+import {concatenateTypedArrays} from '@loaders.gl/loader-utils';
+
+/**
+ * Add skirt to existing mesh
+ * @param {object} attributes - POSITION and TEXCOOD_0 attributes data
+ * @param {any} triangles - indices array of the mesh geometry
```
Comment (kylebarron) [inline modules/terrain/src/lib/helpers/skirt.js:6]: Should we declare this type as `TypedArray | number[]`?

File: modules/terrain/src/lib/parse-quantized-mesh.js [hunk 5]
```diff
@@ -43,9 +44,23 @@ function getTileMesh(arrayBuffer, options) {
   }
   const {bounds} = options;
   // Don't parse edge indices or format extensions
-  const {header, vertexData, triangleIndices} = decode(arrayBuffer, DECODING_STEPS.triangleIndices);
+  const {
+    header,
+    vertexData,
+    triangleIndices: originalTriangleIndices
+  } = decode(arrayBuffer, DECODING_STEPS.triangleIndices);
```
Comment (kylebarron) [inline modules/terrain/src/lib/parse-quantized-mesh.js:51]:
```suggestion
    westIndices,
    northIndices,
    eastIndices,
    southIndices
  } = decode(arrayBuffer, DECODING_STEPS.edgeIndices);
```
This is where you get the edge indices from the quantized mesh https://github.com/visgl/loaders.gl/blob/vb/terrain-skirting/modules/terrain/src/lib/decode-quantized-mesh.js

File: modules/terrain/src/lib/parse-quantized-mesh.js [hunk 6]
```diff
@@ -43,9 +44,23 @@ function getTileMesh(arrayBuffer, options) {
...
+  let triangleIndices = originalTriangleIndices;
   // TODO: use skirt information from file
```
Comment (kylebarron) [inline modules/terrain/src/lib/parse-quantized-mesh.js:53]: Take out todo once you use the `edgeIndices`

**Review by kylebarron — APPROVED** [review]

File: modules/terrain/src/lib/helpers/skirt.js (hunk 3 above)
Comment (belom88) [reply modules/terrain/src/lib/helpers/skirt.js:11]: Where can I take relevant quantized dataset to test it?

Comment (kylebarron) [reply modules/terrain/src/lib/helpers/skirt.js:11]: Maybe simplest to take either of the `.terrain` files in the test suite: https://github.com/visgl/loaders.gl/tree/master/modules/terrain/test/data

Alternatively if you want to test with multiple neighboring tiles in a region, you can generate some from Python. Would it be helpful for me to generate a handful of tiles for you, and I can share a zip file here?

Comment (belom88) [reply modules/terrain/src/lib/helpers/skirt.js:11]: I'll use test on for now.

File: modules/terrain/src/lib/helpers/skirt.js (hunk 4 above)
Comment (belom88) [reply modules/terrain/src/lib/helpers/skirt.js:6]: It raises error in `new triangles.constructor` "This expression is not constructable.
  Type 'Function' has no construct signatures.ts(2351)"

**Comment by belom88** [comment]
Comments are fixed in https://github.com/visgl/loaders.gl/pull/1669

**Pull request merged**